### ISSUE TITLE
Better type safety with enums

### ISSUE THREAD

**Issue by ryan10132** [vojtechhabarta/typescript-generator#33]
Loving the typescript 1.8 enums, gets me much better type safety while still having the wire format of the values be strings.

However I think we can do even better with type safety. Consider the following case, where the enum "Foo" was generated

```
type Foo = "a" | "b" | "c";

// ...

var x: Foo = "a"; // This is type safe
var xIsA = x === "a"; // This is not type safe
var xIsD = x === "d"; // This compiles, which is not ideal
```

Here is my idea:

```
type Foo = "a" | "b" | "c";
const FooValues {
    a: <Foo> "a",
    b: <Foo> "b",
    c: <Foo> "c"
}

// ...

var x: Foo = FooValues.a; // This is type safe
var xIsA = x === FooValues.a; // This is type safe
var xIsD = x === FooValues.d; // This does not compile, which is ideal
```

Writing such a "Values" const alongside every enum manually would be very tedious, but since it's all autogenerated, I think it's fine. Could also be behind a flag for people that don't want it.

What do you think? It would be very straightforward to implement, I'd be happy to submit a PR, would you accept it?

**Comment by vojtechhabarta** [comment]
@ryan10132, having constants for enum values is nice however I have few comments.

#### Extension for implementation files
This only make sense in implementation files (.ts) and not in declaration files (.d.ts) since it generates JavaScript runtime code. So we can implement it as an extension, for example `cz.habarta.typescript.generator.ext.EnumConstantsExtension`.

#### Type safety
Assignment is type safe but comparison is not.
``` typescript
var xIsD = x === Foo.d; // This does not compile, which is ideal
// but
var xIsD = x === "d"; // This still compiles
var xIsD = x === Bar.d; // This also compiles (if Bar.d exists)
```

#### Names of const variable
We can even use the same name for `const` values variable.
``` typescript
type Foo = "a" | "b" | "c";
const Foo = {
    a: <Foo>"a",
    b: <Foo>"b",
    c: <Foo>"c"
}
```
First `Foo` is name of the type while second `Foo` is name of variable so they do not conflict because they are in different space.

#### Names of values
If value is for example `"a-a"` it cannot be used as value name, it is not a valid identifier. Currently this can't happen since value names are taken directly from Java enum names but for example Jackson has different options how to serialize enums so eventually it will happen.

#### Formatting
Try to be consistent with TypeScript formatter, there is no space in `<Foo>"a"`.

#### Overall I think this would be nice extension so implement it if you want :smile:

**Comment by ryan10132** [comment]
Awesome feedback. I'm especially happy that we can use the same name as the type, I'll probably have this implemented over the next few days. Thanks!

**Comment by vojtechhabarta** [comment]
Implemented in #34.
